### ISSUE TITLE
could you provide the scripts of training to reproduce your results? Simply run the default parameters get inferior results. Thanks

### ISSUE THREAD

**Issue by kailigo** [daizuozhuo/batch-feature-erasing-network#2]

**Comment by daizuozhuo** [comment]
Thank you for trying my code. I don't have such a training script since the training schedule depends on the datasets. Usually you should first use Adam optimizer with learning rate 1e-3 for the first 150 epochs. If you encounter NaN loss error, then try to decrease the learning rate. The training program will report the testing accuracy every 20 epochs and save the best model. Then you should change the learning rate to 1e-4 for next 200 epochs and continue training based on the best model from previous step. And finally use 1e-5 learning rate util get the best results. You can also try SGD optimizer with learning rate 1e-4 and 1e-5 for more stable training, but it needs more epochs to converge. The reported accuracy is the best result we got.

**Comment by ningxia** [comment]
Thank you first very much for sharing your code. I downloaded your market model (market_944.pth.tar) and tried to replicate your results by setting the max epochs to 1000 for training (2 x Titan Xp, not as your 4 x 1080), but the evaluations of the first 60 epochs were not very good with losses lingering around 14.631:
mAP: 0.1%
CMC curve
Rank-1  : 0.4%
Rank-2  : 0.5%
Rank-4  : 0.8%
Rank-5  : 0.9%
Rank-8  : 1.2%
Rank-10 : 1.5%
Rank-16 : 2.2%
Rank-20 : 2.4%
--------------

@40th epochs:
mAP: 0.0%
CMC curve
Rank-1  : 0.0%
Rank-2  : 0.0%
Rank-4  : 0.1%
Rank-5  : 0.2%
Rank-8  : 0.4%
Rank-10 : 0.7%
Rank-16 : 1.4%
Rank-20 : 1.9%
--------------

@60 epochs:
mAP: 0.0%
CMC curve
Rank-1  : 0.1%
Rank-2  : 0.2%
Rank-4  : 0.5%
Rank-5  : 0.5%
Rank-8  : 0.9%
Rank-10 : 1.1%
Rank-16 : 1.7%
Rank-20 : 2.2%

When I set the max epochs to 500, I got results on the 20th epoch:
mAP: 55.0%
CMC curve
Rank-1  : 77.9%
Rank-2  : 84.1%
Rank-4  : 89.3%
Rank-5  : 90.6%
Rank-8  : 92.6%
Rank-10 : 93.7%
Rank-16 : 95.5%
Rank-20 : 96.1%
--------------

Is this normal? Can you please provide some guidance? Thanks again.

> Thank you for trying my code. I don't have such a training script since the training schedule depends on the datasets. Usually you should first use Adam optimizer with learning rate 1e-3 for the first 150 epochs. If you encounter NaN loss error, then try to decrease the learning rate. The training program will report the testing accuracy every 20 epochs and save the best model. Then you should change the learning rate to 1e-4 for next 200 epochs and continue training based on the best model from previous step. And finally use 1e-5 learning rate util get the best results. You can also try SGD optimizer with learning rate 1e-4 and 1e-5 for more stable training, but it needs more epochs to converge. The reported accuracy is the best result we got.

**Comment by ningxia** [comment]
@daizuozhuo Thank you first very much for sharing your code. I downloaded your market model (market_944.pth.tar) and tried to replicate your results by setting the max epochs to 1000 for training (2 x Titan Xp, not as your 4 x 1080), but the evaluations of the first 60 epochs were not very good with losses lingering around 14.631:

@20th epochs:
mAP: 0.1%
CMC curve
Rank-1  : 0.4%
Rank-2  : 0.5%
Rank-4  : 0.8%
Rank-5  : 0.9%
Rank-8  : 1.2%
Rank-10 : 1.5%
Rank-16 : 2.2%
Rank-20 : 2.4%

@40th epochs:
mAP: 0.0%
CMC curve
Rank-1  : 0.0%
Rank-2  : 0.0%
Rank-4  : 0.1%
Rank-5  : 0.2%
Rank-8  : 0.4%
Rank-10 : 0.7%
Rank-16 : 1.4%
Rank-20 : 1.9%

@60th epochs:
mAP: 0.0%
CMC curve
Rank-1  : 0.1%
Rank-2  : 0.2%
Rank-4  : 0.5%
Rank-5  : 0.5%
Rank-8  : 0.9%
Rank-10 : 1.1%
Rank-16 : 1.7%
Rank-20 : 2.2%

When I set the max epochs to 500, I got results on the 20th epoch:
mAP: 55.0%
CMC curve
Rank-1  : 77.9%
Rank-2  : 84.1%
Rank-4  : 89.3%
Rank-5  : 90.6%
Rank-8  : 92.6%
Rank-10 : 93.7%
Rank-16 : 95.5%
Rank-20 : 96.1%

Is this normal? Can you please provide some guidance? Thanks again.

**Comment by daizuozhuo** [comment]
@ningxia This is not normal. If you use less GPUs, then you should decrease the learning rate. Maybe you should start training with learning rate 1e-4.

**Comment by ningxia** [comment]
@daizuozhuo I will try smaller learning rate, thanks!

**Comment by ningxia** [comment]
@daizuozhuo Can you please provide your config for obtaining the best market1501 (market_944.pth.tar) model? Thanks.

**Comment by daizuozhuo** [comment]
@ningxia I will release the script when I have spare time. Please be patient.

**Comment by ningxia** [comment]
@daizuozhuo Thanks.

**Comment by daizuozhuo** [comment]
@kailigo @ningxia Hi, the training command is released. Please have a try!

**Comment by ningxia** [comment]
@daizuozhuo I am running it with 2 GPUs, thanks a lot!

**Comment by ningxia** [comment]
@daizuozhuo 
I was running your train script with 2 x Titan Xp, PyTorch 0.4.1, CUDA 9.0, cuDNN 7.0 and got the following results. Which PyTorch, CUDA, cuDNN versions did you use? Thanks.
Epoch: 400
mAP: 84.3%
CMC curve
Rank-1  : 93.3%
Rank-2  : 95.7%
Rank-4  : 97.6%
Rank-5  : 98.0%
Rank-8  : 98.7%
Rank-10 : 98.9%
Rank-16 : 99.3%
Rank-20 : 99.3%

Best rank-1 93.4%, achieved at epoch 360
mAP: 84.1%
CMC curve
Rank-1  : 93.4%
Rank-2  : 95.7%
Rank-4  : 97.7%
Rank-5  : 98.0%
Rank-8  : 98.7%
Rank-10 : 99.0%
Rank-16 : 99.2%
Rank-20 : 99.3%

**Comment by daizuozhuo** [comment]
@ningxia We are using the same software version. Have you update the code? Could you share your training log and command here? If you changed the training batch size, then you should change the training schedule too.

**Comment by ningxia** [comment]
@daizuozhuo Here is the command, the only difference from the one you provided is the number of GPUs:
python3 main_reid.py train --save_dir='./pytorch-ckpt/market-bfe' --max_epoch=400 --eval_step=30 --dataset=market1501 --test_batch=128 --train_batch=128 --optim=adam --adjust_lr --num_gpus 2

[log_train.txt](https://github.com/daizuozhuo/batch-feature-erasing-network/files/2810550/log_train.txt)

**Comment by joberzheng** [comment]
@daizuozhuo   thank you for  providing   train script  ,I try "random erasing" to data augmentation and  erased height ration 0.2 ,Achieving results of mAp 86.7% and rank-1 94.9% , @ningxia  you can try it ,,I will try other config to  obtaining the best  result

**Comment by ningxia** [comment]
@xinxiangzheng 
How many GPUs did you use to obtain that result? Software versions?
Did you mean by changing the default TranTransform's RandomErasing probability to 0.2 or the BatchCrop's height_ratio in BFE to 0.2? Thanks.

**Comment by joberzheng** [comment]
I am training it with 4 GPUs,  PyTorch 0.4.1, CUDA 9.0, cuDNN 7.1.2 ,I change the BatchCrop's height_ratio in BFE to 0.2, train batch 128  ,daizuozhuo apply "cutout" to  data augmentation but  I  use  the default TranTransform's RandomErasing

**Comment by ningxia** [comment]
@xinxiangzheng l believe I was using the same heigh_ratio defined in main_reid.py

elif opt.model_name == 'bfe':
        if opt.datatype == "person":
            model = BFE(dataset.num_train_pids, 1.0, 0.2)

**Comment by joberzheng** [comment]
yes

**Comment by ningxia** [comment]
@xinxiangzheng @daizuozhuo I guess I missed out the "cutout" part of TrainTransform. With it, I got the following results:
On 2 x Titan Xp server:
mAP: 85.3%
CMC curve
Rank-1  : 94.2%
Rank-2  : 96.6%
Rank-4  : 97.6%
Rank-5  : 97.9%
Rank-8  : 98.5%
Rank-10 : 98.8%
Rank-16 : 99.0%
Rank-20 : 99.3%

On 4 x1080Ti server, CUDA 9.0, cuDNN 7.1.2: 
**mAP: 86.2%**
CMC curve
**Rank-1  : 95.2%**
Rank-2  : 96.6%
Rank-4  : 97.8%
Rank-5  : 98.2%
Rank-8  : 98.7%
Rank-10 : 98.9%
Rank-16 : 99.3%
Rank-20 : 99.4%

**Comment by daizuozhuo** [comment]
@ningxia Congratulations! Then I would close this issue.

**Comment by swg209** [comment]
@ningxia @kailigo   With one GPU (GTX 2080),    how to decide learning rate and batch_size  to  get good result as paper?  Thx

**Comment by sksksk1748** [comment]
@daizuozhuo  @ningxia  @kailigo  @swg209  @xinxiangzheng 
Hello, Validation dataset are need ?
Thanks!!
